cpp/src/tsne/tsne_runner.cuh

```diff
@@ -78,6 +88,67 @@ class TSNE_runner {
       CUML_LOG_WARN(
         "# of Nearest Neighbors should be at least 3 * perplexity. Your results"
         " might be a bit strange...");
+
+    auto stream         = handle.get_stream();
+    const value_idx dim = params.dim;
+
+    if (params.init == TSNE_INIT::RANDOM) {
+      random_vector(Y, -0.0001f, 0.0001f, n * dim, stream, params.random_state);
+    } else if (params.init == TSNE_INIT::PCA) {
+      auto components          = raft::make_device_matrix<float>(handle, p, dim);
+      auto explained_var       = raft::make_device_vector<float>(handle, dim);
+      auto explained_var_ratio = raft::make_device_vector<float>(handle, dim);
+      auto singular_vals       = raft::make_device_vector<float>(handle, dim);
+      auto mu                  = raft::make_device_vector<float>(handle, p);
+      auto noise_vars          = raft::make_device_scalar<float>(handle, 0);
+
+      paramsPCA prms;
+      prms.n_cols       = p;
+      prms.n_rows       = n;
+      prms.n_components = dim;
+      prms.whiten       = false;
+      prms.algorithm    = solver::COV_EIG_DQ;
+
+      if constexpr (!is_instance_of<tsne_input, manifold_dense_inputs_t>) {
+        throw std::runtime_error("The tsne_input must be of type manifold_dense_inputs_t");
+      } else {
+        pcaFitTransform(handle,
+                        input.X,
+                        Y,
+                        components.data_handle(),
+                        explained_var.data_handle(),
+                        explained_var_ratio.data_handle(),
+                        singular_vals.data_handle(),
+                        mu.data_handle(),
+                        noise_vars.data_handle(),
+                        prms,
+                        stream);
+
+        auto mean_result = raft::make_device_vector<float, int>(handle, dim);
+        auto std_result  = raft::make_device_vector<float, int>(handle, dim);
+        std::vector<float> h_std_result(dim);
+        const float multiplier = 1e-4;
+
+        auto Y_view       = raft::make_device_matrix_view<float, int>(Y, n, dim);
+        auto Y_view_const = raft::make_device_matrix_view<const float, int>(Y, n, dim);
+
+        auto mean_result_view       = mean_result.view();
+        auto mean_result_view_const = raft::make_const_mdspan(mean_result.view());
+
+        auto std_result_view = std_result.view();
+
+        auto h_multiplier_view_const = raft::make_host_scalar_view<const float>(&multiplier);
+        auto h_std_result_view_const = raft::make_host_scalar_view<const float>(&h_std_result[0]);
+
+        raft::stats::mean(handle, Y_view_const, mean_result_view, false);
+        raft::stats::stddev(handle, Y_view_const, mean_result_view_const, std_result_view, false);
+
+        raft::update_host(h_std_result.data(), std_result.data_handle(), dim, stream);
+
+        raft::linalg::divide_scalar(handle, Y_view_const, Y_view, h_std_result_view_const);
```

Comment:
Can `raft::linalg::unary_op` be used to divide? This would avoid the copy to the host memory.

Reply:
`thrust::transform(handle.get_thrust_policy(), Y_view.data_handle(), Y_view.data_handle() + Y_view().size(), [device_scalar = std_result.data_handle()] __device__ (auto y) { y / device_scalar; });

Reply:
Please don't use thrust::transform. Use `raft::map` instead.